packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -868,6 +887,141 @@ class ExactAssetImage extends AssetBundleImageProvider {
+typedef DeferringImageProviderActionCallback = DeferringImageProviderAction Function();
+
+/// An [ImageProvider] that can delay resolving an image using its wrapped
+/// [imageProvider] until an arbitrary condition is met.
+///
+/// For example, the [Image] widget uses this provider to defer loading images
+/// if it is the child of a scrollable that is rapidly scrolling. It cancels
+/// resolving the image if its element is evicted from the tree before scrolling
+/// ends, and otherwise waits for scrolling to slow down before actually
+/// resolving the image. This prevents the framework from over-utilizing
+/// hardware resources for images that will never be shown on screen.
+///
+/// The provider immediately tries to obtain the key for the image. Once the key
+/// is obtained, it does the following in a loop:
+///
+/// 1. Check whether the [ImageCache] has the image available. If so, call
+///    [ImageProvider.resolveForStream] immediately since it will not involve
+///    any additional decoding work than what is already done or scheduled, and
+///    then break the loop.
+///    In particular, this means that an image loaded via [precacheImage] or by
+///    another [ImageProvider] will be available without deferring.
+/// 2. Call the [handleDeferringImageProviderAction] callback. If it is cancel, break the loop
+///    without resolving the image. If it is resolve, break the loop and call
+///    [ImageProvider.resolveForStream] on the wrapped image provider. If it is
+///    defer, re-trigger the callback on the next frame during the
+///    [SchedulerPhase.midFrameMicrotasks].
+@immutable
+class DeferringImageProvider<T> extends ImageProvider<T> {
+  /// Creates a new [DeferringImageProvider].
+  ///
+  /// All parameters must not be null.
+  const DeferringImageProvider({
+    @required this.imageProvider,
+    @required this.handleDeferringImageProviderAction,
+  }) : assert(imageProvider != null),
+       assert(handleDeferringImageProviderAction != null);
+
+  /// The wrapped [ImageProvider] to delegate loading the image and creating
+  /// a key for it.
+  ///
+  /// The image provider's load method may never be called if resolution gets
+  /// cancelled.
+  final ImageProvider<T> imageProvider;
+
+  /// A callback used by [resolveForStream] to determine whether resolution
+  /// should be cancelled, deferred, or executed.
+  ///
+  /// This callback is never called if the [ImageCache] already has the image
+  /// present, such as from [precacheImage] or another provider successfully
+  /// loading it.
+  ///
+  /// This callback is called:
+  ///
+  /// 1. If the [ImageCache] does not contain this image already AND
+  /// 2. The key for the image has been obtained
+  ///
+  /// While it returns [DeferringImageProviderAction.defer], it schedules a new
+  /// call during the [SchedulerPhase.midFrameMicrotasks] of the next frame.
+  final DeferringImageProviderActionCallback handleDeferringImageProviderAction;
```

Comment:
per our style guide this should be called `onSomething`.

Reply:
Done